### ISSUE TITLE
odata `groupBy` causes `subquery uses ungrouped column "__select__.author_id" from outer query`

### ISSUE THREAD

**Issue by hakimio** [cap-js/cds-dbs#708]
### Description of erroneous behaviour 

CDS server crashes with error `TypeError: transformedQuery.SELECT.from.ref is not iterable` when trying to `groupBy` on a child association.

Stacktrace:
```
[cds] - TypeError: transformedQuery.SELECT.from.ref is not iterable
    at expandColumn (node_modules\@cap-js\db-service\lib\cqn4sql.js:790:41)
    at handleExpand (node_modules\@cap-js\db-service\lib\cqn4sql.js:420:40)
    at node_modules\@cap-js\db-service\lib\cqn4sql.js:354:32
    at getTransformedColumns (node_modules\@cap-js\db-service\lib\cqn4sql.js:382:21)
    at transformSelectQuery (node_modules\@cap-js\db-service\lib\cqn4sql.js:156:41)
    at cqn4sql (node_modules\@cap-js\db-service\lib\cqn4sql.js:73:26)
    at PostgresService.cqn4sql (node_modules\@cap-js\db-service\lib\SQLService.js:375:12)
    at PostgresService.cqn2sql (node_modules\@cap-js\db-service\lib\SQLService.js:351:18)
    at PostgresService.onSELECT (node_modules\@cap-js\db-service\lib\SQLService.js:127:39)
    at PostgresService.onSELECT (node_modules\@cap-js\postgres\lib\PostgresService.js:314:18) {
  id: '1043785',
  level: 'ERROR',
  timestamp: 1719308262527
}
```

### Detailed steps to reproduce

Try to run an odata query similar to the following:
```
$apply: filter(foo eq 'bar')/groupby((user/name),aggregate(amount with sum as totalAmount))
```
If you change `groupby((user/name)` with `groupby((user_ID)` it works as expected.

### Details about your project

| Package | Version |
|:------------------|---------------------------------------|
| OData version     | v4                               |
| @cap-js/cds-typer      | 0.22.0                                                     |
| @cap-js/cds-types      | 0.2.0                                                      |
| @cap-js/postgres       | 1.8.0                                                      |
| @sap/cds               | 7.9.2                                                      |
| @sap/cds-compiler      | 4.9.4                                                      |
| @sap/cds-dk            | 7.9.3                                                      |
| @sap/cds-dk (global)   | 7.9.2                                                      |
| @sap/cds-fiori         | 1.2.4                                                      |
| @sap/cds-foss          | 5.0.1                                                      |
| @sap/cds-mtxs          | 1.18.1                                                     |
| @sap/eslint-plugin-cds | 3.0.4                                                      |
| Node.js                | v20.13.1                                                   |

**Comment by patricebender** [comment]
Hi, could you please let me know which version of the db-service you are using (its not part of the version table you've provided)

**Comment by hakimio** [comment]
My `db-service` is at `v1.10.1`.
BTW, just checked capire and this query should work correctly: https://cap.cloud.sap/docs/advanced/odata#example-1

**Comment by hakimio** [comment]
Also, is odata v4 `filter in` implemented? Doesn't seem to work for me either.
My CDS config:
```json
    "features": {
        "assert_integrity": "db",
        "odata_new_parser": true,
        "odata_new_adapter": false,
        "string_decimals": false
    }
```

**Comment by patricebender** [comment]
> My `db-service` is at `v1.10.1`. BTW, just checked capire and this query should work correctly: https://cap.cloud.sap/docs/advanced/odata#example-1

yes, this should be supported. I have identified your issue and prepared a fix in #709 
Once this change is merged, I will ship a release.

Thank you for reporting this issue 😊

> Also, is odata v4 `filter in` implemented? Doesn't seem to work for me either. My CDS config:
> 
> ```json
>     "features": {
>         "assert_integrity": "db",
>         "odata_new_parser": true,
>         "odata_new_adapter": false,
>         "string_decimals": false
>     }
> ```

Sorry, I can't help you as I'm no expert with the ODATA adapter. Maybe @johannes-vogel can help you out.

**Comment by patricebender** [comment]
I was able to reproduce your issue. The problem is - as the error states - that the group by clause of the outer query is missing the `author.ID` which is used to build the correlated subquery in for the expand.

Long story short, I was able to reproduce your issue, in my bookshop.
1. `http://localhost:4004/admin/Books?$apply=filter(title%20ne%20%27bar%27)/groupby((author/name))`
2. observe the error: `subquery uses ungrouped column "__select__.author_id" from outer query`

subquery:
```sql
SELECT author2.name
FROM AdminService_Authors as author2
WHERE __select__.author_ID = author2.ID
LIMIT $1
```

main query `__select__`:
```sql
SELECT (
-- expand subquery from above
) as author
FROM (
   …
) as __select__
left JOIN …
GROUP BY author.name -- this is missing the author.ID
```

For me, including the `author/ID` in my odata request, did the trick

```diff
- http://localhost:4004/admin/Books?$apply=filter(title%20ne%20%27bar%27)/groupby((author/name))
+ http://localhost:4004/admin/Books?$apply=filter(title%20ne%20%27bar%27)/groupby((author/name,author/ID))
```

```json
{
  "value": [
    {
      "author": {
        "ID": 150,
        "name": "Edgar Allen Poe"
      }
    },
    {
      "author": {
        "ID": 107,
        "name": "Charlotte Brontë"
      }
    },
    {
      "author": {
        "ID": 101,
        "name": "Emily Brontë"
      }
    },
    {
      "author": {
        "ID": 170,
        "name": "Richard Carpenter"
      }
    }
  ],
  "@odata.context": "$metadata#Books(author(name,ID))"
}
```

Could you try that? I will ask my colleagues if we can give you some convenience for this in the future, as this is a technical detail which one might not expect.

**Comment by hakimio** [comment]
While yes, including `author/ID`, is a valid workaround which works for me as well, `groupby(author/name)` should also work since it is documented in the [capire docs](https://cap.cloud.sap/docs/advanced/odata#example-1) and [odata docs](https://devblogs.microsoft.com/odata/aggregation-extensions-in-odata-asp-net-core/#groupby-transformation).
In this particular case `groupby(author/name)` and `groupby(author/name, author/ID)` are equivalent but there are cases when you want to group by attributes which will not have the equivalent aggregation result.

**Comment by patricebender** [comment]
Of course, you are correct. I tried the same on SQlite and there doesn't seem to be a problem. I will bring this to the team and we will discuss this. Hope you can continue with the workaround, for now.

**Comment by hakimio** [comment]
Here is another `groupby` query which results in a different error:

**Query:**
```
$apply=filter(title ne 'bar')/groupby((author/country/name, author/country/ID))
```

**Error:**
```
[cds] - Error: An association can't be used as a value in an expression
    at rejectAssocInExpression (node_modules\@cap-js\db-service\lib\cqn4sql.js:1540:13)
    at assertNoStructInXpr (node_modules\@cap-js\db-service\lib\cqn4sql.js:1534:7)
    at getTransformedTokenStream (node_modules\@cap-js\db-service\lib\cqn4sql.js:1396:11)
    at cqn4sql (node_modules\@cap-js\db-service\lib\cqn4sql.js:64:31)
    at transformSubquery (node_modules\@cap-js\db-service\lib\cqn4sql.js:969:12)
    at getTransformedFrom (node_modules\@cap-js\db-service\lib\cqn4sql.js:1578:25)
    at cqn4sql (node_modules\@cap-js\db-service\lib\cqn4sql.js:69:51)
    at PostgresService.cqn4sql (node_modules\@cap-js\db-service\lib\SQLService.js:375:12)
    at PostgresService.cqn2sql (node_modules\@cap-js\db-service\lib\SQLService.js:351:18)
    at PostgresService.onSELECT (node_modules\@cap-js\db-service\lib\SQLService.js:127:39) {
  id: '1290165',
  level: 'ERROR',
  timestamp: 1719421678688
}
```

**Comment by patricebender** [comment]
> Here is another `groupby` query which results in a different error:
> 
> **Query:**
> 
> ```
> $apply=filter(title ne 'bar')/groupby((author/country/name, author/country/ID))
> ```
> 
> **Error:**
> 
> ```
> [cds] - Error: An association can't be used as a value in an expression
>     at rejectAssocInExpression (node_modules\@cap-js\db-service\lib\cqn4sql.js:1540:13)
>     at assertNoStructInXpr (node_modules\@cap-js\db-service\lib\cqn4sql.js:1534:7)
>     at getTransformedTokenStream (node_modules\@cap-js\db-service\lib\cqn4sql.js:1396:11)
>     at cqn4sql (node_modules\@cap-js\db-service\lib\cqn4sql.js:64:31)
>     at transformSubquery (node_modules\@cap-js\db-service\lib\cqn4sql.js:969:12)
>     at getTransformedFrom (node_modules\@cap-js\db-service\lib\cqn4sql.js:1578:25)
>     at cqn4sql (node_modules\@cap-js\db-service\lib\cqn4sql.js:69:51)
>     at PostgresService.cqn4sql (node_modules\@cap-js\db-service\lib\SQLService.js:375:12)
>     at PostgresService.cqn2sql (node_modules\@cap-js\db-service\lib\SQLService.js:351:18)
>     at PostgresService.onSELECT (node_modules\@cap-js\db-service\lib\SQLService.js:127:39) {
>   id: '1290165',
>   level: 'ERROR',
>   timestamp: 1719421678688
> }
> ```

I was not able to reproduce this specific scenario (can you provide a sample?). However, with the same query I was running in a different error which is fixed by #715, I will continue investigating in keep you updated. I'm afraid, for the postgres issue, we do not have a quick fix available. I already found out that SQlite can work with the exact same query. Once #712 is fixed, I will also give this a try on HANA.

**Comment by hakimio** [comment]
I'll just wait for the fix in #715 to land and if the error still persists, I'll see if I can make a better reproduction.

**Comment by hakimio** [comment]
@patricebender a small update: the previous error was unrelated to `groupby`, it was caused by a filter similar to `filter(author ne null)` which, based on the error, I assume is invalid. 

Coming back to `groupby` issue. `groupby((author/country/name, author/country/ID))` throws an error, but `groupby((author/ID, author/country/name, author/country/ID))` works as intended.

On slightly related note, can you tell me if the `totalValue` calculated column is valid in the following `cds` schema:
```cds
entity Customers : cuid, managed {
    contracts       : Composition of many Contracts
                          on contracts.customer = $self;
    totalValue      : Decimal(15, 2) = sum(
        contracts[status = 'Active' and orderProduct.product.isRecurring = true].orderProduct.price
    );
}

entity Contracts : cuid, managed {
    status               : ContractStatus;
    orderProduct         : Association to OrderProducts;
}

entity OrderProducts : cuid, managed {
    product           : Association to Products;
}

entity Products : cuid, managed {
    isRecurring                : Boolean default true;
}
```
No errors shown by the SAP language service, but trying to generate migration with `cds deploy --dry --delta-from cds-model.csn > delta.sql` throws an error:
```
crm-service.cds:14:44: Error: “CrmService.Customers”: “product” is not foreign key of managed association “orderProduct” in path “orderProduct.product.isRecurring” (in entity:“CrmService.Customers”/query:1)
crm-service.cds:14:44: Error: “CrmService.Customers”: “isRecurring” is not foreign key of managed association “product” in path “orderProduct.product.isRecurring” (in entity:“CrmService.Customers”/query:1)
    at throwWithError (node_modules\@sap\cds-compiler\lib\base\messages.js:559:13)
    at translateAssocsToJoins (node_modules\@sap\cds-compiler\lib\transform\translateAssocsToJoins.js:72:3)
    at translateAssocsToJoinsCSN (node_modules\@sap\cds-compiler\lib\transform\translateAssocsToJoins.js:26:3)
    at handleAssocToJoins (node_modules\@sap\cds-compiler\lib\transform\forRelationalDB.js:694:20)
    at Proxy.transformForRelationalDBWithCsn (node_modules\@sap\cds-compiler\lib\transform\forRelationalDB.js:227:5)
    at csnForSql (node_modules\@sap\cds-compiler\lib\api\main.js:209:42)
    at sqlMigration (node_modules\@sap\cds-compiler\lib\api\main.js:476:20)
    at Function.api [as migration] (node_modules\@sap\cds-compiler\lib\api\main.js:1127:22)
    at Function.migration (node_modules\@sap\cds-compiler\lib\main.js:142:41)
    at Object.deltaSql (node_modules\@sap\cds\lib\compile\cdsc.js:162:53)
```

**Comment by patricebender** [comment]
compiling your model with `cds compile -2 sql` does not work:

```
❯ cds compile -2 sql e.cds
[ERROR] e.cds:35:13: “V”: “isRecurring” is not foreign key of managed association “product” in path “orderProduct.product.isRecurring” (in entity:“V”/query:1)
[ERROR] e.cds:35:13: “V”: “product” is not foreign key of managed association “orderProduct” in path “orderProduct.product.isRecurring” (in entity:“V”/query:1)
```

---

The calculated element is not valid. That is because it navigates to a non-foreign-key in path `orderProduct.product.isRecurring` where the association `product` is not fk of `orderProduct` and `isRecurring` is not a fk element of `product`.

This is a known limitation. The filter path along `contracts[…].orderProduct` induces a join by itself, with `[…]` being part of the `join` node.

this would be the resulting view if each element along the path in question would be a foreign key element:
```sql
CREATE VIEW V AS
SELECT Customers_0.ID,
  sum(orderProduct_2.price) AS totalValue
FROM (
    (
      Customers AS Customers_0
      LEFT JOIN Contracts AS contracts_1 ON ((contracts_1.customer_ID = Customers_0.ID))
      AND (
        contracts_1.status = 'Active'
        AND contracts_1.orderProduct_product_isRecurring = TRUE
      )
    )
    LEFT JOIN OrderProducts AS orderProduct_2 ON (contracts_1.orderProduct_ID = orderProduct_2.ID)
    AND (
      contracts_1.orderProduct_product_ID = orderProduct_2.product_ID
    )
    AND (
      contracts_1.orderProduct_product_isRecurring = orderProduct_2.product_isRecurring
    )
  )
```

→ The filter path is part of the join. If the filter condition itself induces joins (due to non-foreign-key relations), things get much more complex.

**Comment by hakimio** [comment]
I see. Thank you for the detailed explanation.

**Comment by hakimio** [comment]
@patricebender any idea when the new version will be released?

**Comment by patricebender** [comment]
> @patricebender any idea when the new version will be released?

Hi @hakimio,

we released https://github.com/cap-js/cds-dbs/releases/tag/db-service-v1.10.3 on friday :)

BR
Patrice

**Comment by hakimio** [comment]
@patricebender a couple of issues:
- while the initial problem is fixed and `groupby(author/name)` now works, query with multiple `groupby` columns like `groupby(author/name, author/ID)` doesn't seem to work anymore. Is that intended?
- given the following valid cds schema:
```cds
entity Customers : cuid, managed {
    activeContracts : Integer = count(
        contracts[status = 'Active'].ID
    );
}
entity Contracts : cuid, managed {
    status               : ContractStatus;
}

service CrmService @(path: '/api') {
    entity Customers                    as projection on Customers;
    entity Contracts                    as projection on Contracts;
}
```

`cds compile -2 sql` generates invalid `SQL` like the following:
```sql
CREATE VIEW CrmService_Customers AS SELECT
  Customers_0.ID,
  Customers_0.createdAt,
  Customers_0.createdBy,
  Customers_0.modifiedAt,
  Customers_0.modifiedBy,
  count(contracts_1.ID) AS activeContracts
FROM (gc_crm_Customers AS Customers_0 LEFT JOIN gc_crm_Contracts AS contracts_1 ON ((contracts_1.customer_ID = Customers_0.ID)) AND (contracts_1.status = 'Active'));
```

which results in `SQL` error:
```
column "customers_0.id" must appear in the GROUP BY clause or be used in an aggregate function
```

**Comment by patricebender** [comment]
> query with multiple groupby columns like groupby(author/name, author/ID) doesn't seem to work anymore. Is that intended?

could you elaborate on that? What exactly does not work?

> cds compile -2 sql generates invalid SQL like the following:

Even though you are correct and the sql is not valid, you are responsible for providing a proper group by clause here → group the results by the non-aggregated columns.

The compiler is not aware of the aggregation nature of the `count` function. The `cds-compiler` will likely not implement a semantic check here and will not automagically add the correct group by for you, I'm afraid.

**Comment by hakimio** [comment]
> could you elaborate on that? What exactly does not work?

`groupby(author/name, author/ID)` query now produces error (used to work before):
```
[cds] - error: subquery uses ungrouped column "user.id" from outer query
    at node_modules\pg\lib\client.js:526:17
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.all (node_modules\@cap-js\postgres\lib\PostgresService.js:174:26)
    at async PostgresService.onSELECT (node_modules\@cap-js\db-service\lib\SQLService.js:132:16)
    at async next (node_modules\@sap\cds\lib\srv\srv-dispatch.js:68:17)
    at async PostgresService.handle (node_modules\@sap\cds\lib\srv\srv-dispatch.js:66:10)
    at async ApplicationService.<anonymous> (node_modules\@sap\cds\libx\_runtime\common\generic\crud.js:73:16)
    at async next (node_modules\@sap\cds\lib\srv\srv-dispatch.js:68:17)
    at async ApplicationService.handle (node_modules\@sap\cds\lib\srv\srv-dispatch.js:66:10)
    at async ApplicationService.handle (node_modules\@sap\cds\libx\_runtime\common\Service.js:87:16) {
  length: 141,
  severity: 'ERROR',
  code: '42803',
  detail: undefined,
  hint: undefined,
  position: '242',
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'parse_agg.c',
  line: '1450',
  routine: 'check_ungrouped_columns_walker',
  query: 'SELECT to_jsonb(root.*) as _json_ FROM (SELECT jsonb(root."user") as "user",root.totalAmount as "totalAmount" FROM (SELECT (SELECT to_jsonb("false".*) as _json_ FROM (SELECT "false".name as "name","false".ID as "ID" FROM (SELECT "user".name,"user".ID LIMIT $1) as "false") as "false") as "user",sum(__select__.amount) as totalAmount FROM (SELECT TimeTransactions.ID,TimeTransactions.oldCrmId,TimeTransactions.createdAt,TimeTransactions.createdBy,TimeTransactions.modifiedAt,TimeTransactions.modifiedBy,TimeTransactions.timeVault_ID,TimeTransactions.amount,TimeTransactions.user_ID,TimeTransactions.comment_ID,TimeTransactions.start,TimeTransactions."end",TimeTransactions.type,TimeTransactions.topic_ID FROM CrmService_TimeTransactions as TimeTransactions left JOIN CrmService_TimeVaults as timeVault ON timeVault.ID = TimeTransactions.timeVault_ID WHERE TimeTransactions.start >= $2 and TimeTransactions.start <= $3 and ((timeVault.entity = $4 and timeVault.entityId = $5) or (timeVault.entity = $6 and (timeVault.entityId = $7 or timeVault.entityId = $8 or timeVault.entityId = $9 or timeVault.entityId = $10 or timeVault.entityId = $11 or timeVault.entityId = $12 or timeVault.entityId = $13 or timeVault.entityId = $14 or timeVault.entityId = $15 or timeVault.entityId = $16 or timeVault.entityId = $17 or timeVault.entityId = $18 or timeVault.entityId = $19 or timeVault.entityId = $20 or timeVault.entityId = $21 or timeVault.entityId = $22 or timeVault.entityId = $23 or timeVault.entityId = $24 or timeVault.entityId = $25 or timeVault.entityId = $26 or timeVault.entityId = $27 or timeVault.entityId = $28 or timeVault.entityId = $29))) ORDER BY TimeTransactions.ID ASC) as __select__ left JOIN CrmService_Users as "user" ON "user".ID = __select__.user_ID GROUP BY "user".name,__select__.user_ID LIMIT $30) as root) as root\n' +
    ' ^',
  id: '1174385',
  level: 'ERROR',
  timestamp: 1720433033312
}
```

> Even though you are correct and the sql is not valid, you are responsible for providing a proper group by clause here → group the results by the non-aggregated columns.

So the service definition should look like the following or the `group by` should go somewhere else?
```cds
service CrmService @(path: '/api') {
    entity Customers                    as projection on Customers group by ID;
    entity Contracts                    as projection on Contracts;
}
```

**Comment by patricebender** [comment]
> groupby(author/name, author/ID) query now produces error (used to work before):

💯 thank you once again.
→ It looks like we didn't do the necessary foreign key optimization in this situation. It should work with #734

> So the service definition should look like the following or the group by should go somewhere else?

exactly. It is a little cumbersome in this situation, as you also need to add the elements coming in from the `managed` aspect. Maybe in your situation it would be easier to not do a wildcard select for your projection but to explicitly select what you need. E.g.:

```cds
    entity Customers                    as projection on Customers excluding { activeContracts };
    entity CustomersWithActiveContracts as select from Customers { ID, name, activeContracts} group by ID, name
```

Maybe we could add some support from compiler side so that you dont have to write all the non-aggregates into the group by manually.

**Comment by hakimio** [comment]
> 💯 thank you once again.

Thank you for creating a PR to fix this so fast 🙂

>  It is a little cumbersome in this situation, as you also need to add the elements coming in from the managed aspect.

It is enough to just group by the customer `ID` in this case, I don't have to group by all the selected fields. So, it's not too much of a problem.

**Comment by hakimio** [comment]
@patricebender 
Just tested `v1.11.0`. `groupby(author/name, author/ID)` now works without throwing errors but there is still one more small issue. The author `ID` is no longer present in the server response. I expect both the author `name` and `ID` to be selected in this case.

**Comment by hakimio** [comment]
The entity I tested with does have an `ID` field. Also, your server response looks strange to me as well.

I would expect the response to be:
```json
{
    "author": {
        "name": "Edgar Allen Poe",
        "ID": 150
    }
}
```
instead of:
```json
{
    "author": {
        "name": "Edgar Allen Poe",
        "author_ID": 150
    }
}
```

At least that's how it used to work before.
Can you test with some other entities?

**Comment by patricebender** [comment]
It is expected that the foreign key `author_ID` is returned as this is also the value which we `group by`. The query which is sent to the DB is the following:

```sql
    SELECT (
                        SELECT to_jsonb("false".*) as _json_
                        FROM (
                                SELECT "false".name as "name",
                                    "false".author_ID as "author_ID"
                                FROM (
                                        SELECT author.name,
                                            __select__.author_ID
                                        LIMIT $1
                                    ) as "false"
                            ) as "false"
                    ) as author
                FROM (
                        SELECT Books.createdAt,
                            Books.createdBy,
                            Books.modifiedAt,
                            Books.modifiedBy,
                            Books.ID,
                            Books.title,
                            Books.descr,
                            Books.author_ID,
                            Books.genre_ID,
                            Books.stock,
                            Books.price,
                            Books.currency_code
                        FROM localized_AdminService_Books as Books
                        WHERE Books.title is distinct
                        from $2
                        ORDER BY Books.ID ASC
                    ) as __select__
                    left JOIN AdminService_Authors as author ON author.ID = __select__.author_ID
                GROUP BY author.name,
                    __select__.author_ID -- we group by the foreign key
                LIMIT $3
            ) as root
```

now, if we would use `ID` from the join node `author` we would get the initial error message `subquery uses ungrouped column …` error. That is actually the fixed issue in #734. `author/ID` is always optimized to the foreign key, if possible.

The subquery here, is just selecting the columns from the outer query **which are also present in the group by**. The from is not defined as can be seen here:

```sql
SELECT author.name,
     __select__.author_ID
LIMIT $1
```

I get the same result with an aggregation:

<img width="1289" alt="Screenshot 2024-07-09 at 15 12 54" src="https://github.com/cap-js/cds-dbs/assets/23496018/a7d1bf73-36f3-44ba-bcfd-e576abc8c53f">

**Comment by hakimio** [comment]
Ok, using the latest `db-service` (`1.11.0`) and `postgres` (`1.9.1`) packages. Still not getting any `ID` returned.
My cds config:
```json
{
    "features": {
        "assert_integrity": "db",
        "odata_new_parser": true,
        "odata_new_adapter": false,
        "string_decimals": false
    }
}
```

OData query:
```
TimeTransactions?$apply=filter(user_ID eq dcd388a6-20b7-44b9-a296-d0e7d3628765 and start ge 2024-07-05T00:00:00Z and start le 2024-07-09T23:59:59Z)/groupby((topic/ID,topic/headline),aggregate(amount with sum as totalAmount))
```

Generated SQL:
```sql
SELECT to_jsonb(root.*) AS _json_
FROM
  (SELECT jsonb(root.topic) AS "topic",
          root.totalAmount AS "totalAmount"
   FROM
     (SELECT
        (SELECT to_jsonb("false".*) AS _json_
         FROM
           (SELECT "false".topic_ID AS "topic_ID",
                   "false".headline AS "headline"
            FROM
              (SELECT __select__.topic_ID,
                      topic.headl ine
               LIMIT $1) AS "false") AS "false") AS topic,
             sum(__select__.amount) AS totalAmount
      FROM
        (SELECT TimeTransactions.ID,
                TimeTransactions.oldCrmId,
                TimeTransactions.createdAt,
                TimeTransactions.createdBy,
                TimeTransactions.modifiedAt,
                TimeTransactions.modifiedBy,
                TimeTransactions.timeVault_ID,
                TimeTransactions.amount,
                TimeTransactions.user_ID,
                TimeTransactions.comment_ID,
                TimeTransactions.start,
                TimeTransactions."end",
                TimeTransactions.type,
                TimeTransactions.topic_ID
         FROM CrmService_TimeTransactions AS TimeTransactions
         WHERE TimeTransactions.user_ID = $2
           AND TimeTransactions.start >= $3
           AND TimeTransactions.start <= $4
         ORDER BY TimeTransactions.ID ASC) AS __select__
      LEFT JOIN CrmService_TimeTransactionTopics AS topic ON topic.ID = __select__.topic_ID
      GROUP BY __select__.topic_ID,
               topic.headline
      LIMIT $5) AS root) AS root
```

Server response:
```json
{
    "@odata.context": "$metadata#TimeTransactions(topic(ID,headline),totalAmount)",
    "value": [
        {
            "topic": {
                "headline": "Contract optimization"
            },
            "totalAmount@odata.type": "#Decimal",
            "totalAmount": 820,
            "@odata.id": null
        }
    ]
}
```

Can you spot any differences from your setup?

**Comment by patricebender** [comment]
> Can you spot any differences from your setup?

actually yes, here is my config:

```json
  "features": {
    "odata_new_adapter": true,
    "string_decimals": false
  }
```

unfortunately, with your configuration I always run into

```xml
<error xmlns="http://docs.oasis-open.org/odata/ns/metadata">
<code>400</code>
<message>Expected uri token 'ODataIdentifier' could not be found in 'filter(title ne 'bar')/groupby((author/name, author/ID),aggregate(price with sum as totalAmount))' at position 45</message>
</error>
```

→ the same query returns proper results with my configuration, though.

**Comment by hakimio** [comment]
Ok, it does return the `ID` with the `odata_new_adapter` enabled but then I get a bunch of errors like:
```
Value "2024-07-10T05:46:23.731Z" is invalid according to type definition "cds.DateTime"
```

@johannes-vogel any idea why the old odata adapter doesn't return the `ID`?

**Comment by johannes-vogel** [comment]
The expected OData result is:

```json
{
    "author": {
        "name": "Edgar Allen Poe",
        "ID": 150 // nested "author_ID" is invalid as "author" entity does not have that field.
    }
}
```

If author_ID is provided instead of ID, the old odata adapter cleanses it, I guess.
You could simply check by providing the different result structure in a custom handler.

**Comment by hakimio** [comment]
@johannes-vogel 
So, it's an issue with the `db-service` and not the `odata parser` after all?

Off topic question: does any odata parser support odata `filter in` operation? If not, are there any plans to add support for it?

**Comment by johannes-vogel** [comment]
> @johannes-vogel So, it's an issue with the `db-service` and not the `odata parser` after all?

yes, i do think so.

> Off topic question: does any odata parser support odata [`filter in`](https://learn.microsoft.com/en-us/odata/webapi/in-operator) operation? If not, are there any plans to add support for it?

yes, new odata adapter will support it.

**Comment by patricebender** [comment]
> So, it's an issue with the db-service and not the odata parser after all?

I feel like adding an alias for the foreign key will do the trick. I will prepare a fix. :)

**Comment by patricebender** [comment]
hi @hakimio, we have shipped a fix for your issue. Let me know if it works out for you :)

**Comment by hakimio** [comment]
Hi @patricebender
Sorry, for late response. After upgrading to `db-service v1.12.0` everything seems to be working as it should.
Thank you for the fix 🙂
